raiden/utils/upgrades.py

```diff
@@ -0,0 +1,127 @@
+    """ This class is responsible for figuring out which migrations
+    need to be executed in order to bring the database up to date
+    with the current implementation.
+    """
+    def __init__(self, db_filename: str):
+        self._current_db_filename = Path(db_filename)
+
+    def run(self):
+        """
+        The `_current_db_filename` is going to hold the filename of the database
+        with the new version. However, the previous version's data
+        is going to exist in a file whose name contains the old version.
+        Therefore, running the migration means that we have to copy
+        all data to the current version's database, execute the migration
+        functions.
+        """
+        if get_db_version(self._current_db_filename) == RAIDEN_DB_VERSION:
+            # The current version has already been created / updraded.
+            return
+        else:
+            # The version inside the current database was not the expected one.
+            # Delete and re-run migration
+            self._delete_current_db()
+
+        old_version, old_db_filename = older_db_file(str(self._current_db_filename.parent))
+
+        if not old_version:
+            # There are no older versions to upgrade from.
+            return
+
+        self._copy(str(old_db_filename), str(self._current_db_filename))
+
+        storage = SQLiteStorage(str(self._current_db_filename), JSONSerializer())
+
+        log.debug(f'Upgrading database to v{RAIDEN_DB_VERSION}')
+
+        storage.conn.isolation_level = None
+        cursor = storage.conn.cursor()
+        cursor.execute('BEGIN')
+        try:
+            for upgrade_func in UPGRADES_LIST:
+                upgrade_func(cursor, old_version, RAIDEN_DB_VERSION)
+
+            # COMMIT is going to be executed within update_version.
+            storage.update_version()
+            # Prevent the upgrade from happening on next restart
+            self._backup_old_db(old_db_filename)
+        except RaidenDBUpgradeError as e:
+            cursor.execute('ROLLBACK')
+            self._delete_current_db()
+            log.error(f'Failed to upgrade database: {str(e)}')
+            raise
+
+    def _backup_old_db(self, filename):
+        backup_name = filename.replace('_log.db', '_log.backup')
+        with get_file_lock(filename):
+            shutil.move(filename, backup_name)
+
+    def _delete_current_db(self):
+        os.remove(str(self._current_db_filename))
+
+    def _copy(self, old_db_filename, current_db_filename):
+        with get_file_lock(old_db_filename), get_file_lock(current_db_filename):
+            old_conn = sqlite3.connect(
+                old_db_filename,
+                detect_types=sqlite3.PARSE_DECLTYPES,
+            )
+            current_conn = sqlite3.connect(
+                current_db_filename,
+                detect_types=sqlite3.PARSE_DECLTYPES,
+            )
+
+            sqlitebck.copy(old_conn, current_conn)
+
+            old_conn.close()
```

Comment:
since these connections are not returned, the context manager should be used to also properly handle exceptions:

```python
with closing(old_conn), closing(current_con):
```

https://docs.python.org/3.6/library/contextlib.html#contextlib.closing